parlai/core/torch_agent.py

```diff
@@ -1004,31 +1002,30 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None):
         if optim_states and saved_optim_type != opt['optimizer']:
             # we changed from adam to adamax, or sgd to adam, or similar
             logging.warn('Not loading optim state since optim class changed.')
+            return False
         elif optim_states:
             # check for any fp16/fp32 conversions we need to do
             optimstate_fp16 = 'loss_scaler' in optim_states
             if self.fp16 and optimstate_fp16:
                 # previously trained in fp16, now we're training in fp16.
-                # ideally no action needed, but APEX broke backwards
-                # compatibility and this is the hack around it.
-                optim_states['loss_scaler'] = self.optimizer.state_dict()['loss_scaler']
+                pass
             elif optimstate_fp16 and not self.fp16:
                 # old optimizer was fp16 but now we're doing fp32,
-                # if apex, drop the fp16 wrapper from the state_dict and just load
+                # if pytorch, drop the fp16 wrapper from the state_dict and just load
                 # the fp16 weights into the fp32 tensors
                 if 'optimizer_state_dict' in optim_states:
-                    # trained with apex
+                    # trained with apex, pull in the old state
                     optim_states = optim_states['optimizer_state_dict']
             elif not optimstate_fp16 and self.fp16:
                 # old optimizer was fp32, but now we're doing fp16.
                 # this is a bit clunky, but alternatives are worse
                 try:
-                    self.optimizer.optimizer.load_state_dict(optim_states)
+                    self.optimizer.load_state_dict(optim_states)
                 except ValueError:
                     warn_once(
                         'WARNING: not loading optim state since model params changed.'
                     )
-                return
+                return True
```

Comment:
nit: for clarification, maybe it'd be useful to add in the docstring what a return value of True or False indicates?